File: web/src/list_navigation.ts
import type { ListItemPrimaryRole } from "./protocol";

export type ListNavigationDecision = "down" | "up" | "first" | "last"
  | "pageDown" | "pageUp" | "invokePrimary" | "back";

/** One focused-row keyboard table shared by every DOM Page/List renderer. */
export function listNavigationDecision(
  key: string,
  primaryRole: ListItemPrimaryRole,
): ListNavigationDecision | undefined {
  switch (key) {
    case "Enter": return primaryRole === "static" ? undefined : "invokePrimary";
    case " ": return primaryRole === "toggle" ? "invokePrimary" : "pageDown";
    case "ArrowDown":
    case "j": return "down";
    case "ArrowUp":
    case "k": return "up";
    case "Home":
    case "g": return "first";
    case "End":
    case "G": return "last";
    case "PageDown": return "pageDown";
    case "PageUp": return "pageUp";
    case "Escape":
    case "q": return "back";
    default: return undefined;
  }
}

File: swift/Sources/UnpeelAppKitUI/ListNavigation.swift
import Foundation

/// Platform-neutral keys understood by the shared focused-row decision table.
public enum UIListNavigationKey: Equatable, Sendable {
    case down
    case up
    case first
    case last
    case pageDown
    case pageUp
    case enter
    case space
    case back
}

public enum UIListNavigationDecision: Equatable, Sendable {
    case down
    case up
    case first
    case last
    case pageDown
    case pageUp
    case invokePrimary
    case back
}

/// One keyboard decision table shared by every native Page/List renderer.
/// Routing remains server-driven; `invokePrimary` only asks the caller to emit
/// the action declared by the current authoritative row.
public func uiListNavigationDecision(
    key: UIListNavigationKey,
    primaryRole: UIListItemPrimaryRole
) -> UIListNavigationDecision? {
    switch key {
    case .enter:
        primaryRole == .static ? nil : .invokePrimary
    case .space:
        primaryRole == .toggle ? .invokePrimary : .pageDown
    case .down: .down
    case .up: .up
    case .first: .first
    case .last: .last
    case .pageDown: .pageDown
    case .pageUp: .pageUp
    case .back: .back
    }
}
